Adding support for the new encryption protocol (updated version 2023)

sdb9696 opened the pull request:
I'm not sure if this still welcome as it's been some time since the last attempt to get this working but I've updated the original changes by @SimonWilkinson to hopefully incorporate all the comments on the PR plus a bit more.  Summary of updates:

- Moved klap specific logic out of the Discovery class
- Moved xor specific logic out of the Discovery class
- Introduced an abstraction of an AuthenticationProtocol which the Discovery class is aware of as it needs to be in order to deal with successful vs unsuccessful authentication scenarios
- Introduced a new device type of UnauthenticatedDevice which allows the discovery to show useful information for a device present on the network that fails to authenticate (In practise this is actually happening quite frequently as kasa cloud seems to mess with the device owner periodically and manually switching the device off/on via the app seems to resolve it.
- Moved klap specific logic out of the Auth class and renamed it AuthCredentials
- Updated discover_single to work with the both protocols enabled and if authentication fails to make a discover datagram request to get the unauthenticated device info for just that device
- The changes above will permit any new authenticating protocols that could be introduced by kasa to be added with minimal changes (or none at all) to core classes and discovery logic.
- Updated the encrypt/decrypt logic from Simon to increment the sequence number on each encrypt, which is required to prevent periodic 403 errors.  The klap device is a bit forgiving of sequence numbers but not completely
- Introduced tests for discovery, encryption, authentication failures etc.  Created a fake klap endpoint to facilitate testing
- Added a fixture for HS100(UK) which what my testing took place on.  I also was able to test the old protocol with smart plug devices on my network that did not pick up the new protocol.

This is what the discovery output looks like for an unauthenticated device:

== Unknown - HS100(UK) ==
        Host: 192.168.2.30
        Device state: True
        == Generic information ==
        Time:         None (tz: None
        Hardware:     4.1
        Software:     Unknown
        MAC (rssi):   12:34:56:78:9A:BC (None)
        Location:     None

        == Device specific information ==
        ip: 192.168.3.30
        mac: 12-34-56-78-9A-BC
        device_id: 6BEC350F98852CA1684CF5C7BC6D8A40
        owner: 994661e5222b8e5e3e1d90e73a322315
        device_type: IOT.SMARTPLUGSWITCH
        device_model: HS100(UK)
        hw_ver: 4.1
        factory_default: True
        mgt_encrypt_schm.is_support_https: False
        mgt_encrypt_schm.encrypt_type: KLAP
        mgt_encrypt_schm.http_port: 80
        error_code: 0
        python-kasa.tplinkklap.auth_owner_hash: 9f0b375a2ce82373a53f2a2f222f9368
        python-kasa.tplinkklap.auth_message: The owner hashes do not match, if you expected authentication
                                             to work try switching the device on and off via the Kasa app
                                             to see if the device owner gets corrected.

mystcb commented:
Looks like you beat me to this by about a week! Thanks for getting this updated, and apologies my PR got a little lost (#267). 

Just did a test and it worked with the sockets and setup I have here! Thank you for the additional work you put into this too.

C

sdb9696 commented:
I have updated this to include the 9999 port configurable changes from #471 (still only works with the old protocol atm) and reverted my change of the "target" variable name so it doesn't break homeassistant.  I've now tested this with homeassistant and have some changes shortly to be in a home assistant fork which will enable support for both protocols and unauthenticated devices.

sdb9696 commented:
Hi, I've updated this PR so it passes all the CI tests including pypy-3.8 windows-latest which was intermittently problematic.  There's also a fix for an issue introduced yesterday by a Cython release which affected the pypy-3.8 ubuntu-latest and I think will affect all CI runs regardless of my changes.  @rytilahti I know it's a chunky PR but do you have a feel for when we might be able to start reviewing?

bdraco commented on kasa/protocol.py:
```diff
@@ -25,23 +25,102 @@
+        self.port = port
+        self.timeout = self.DEFAULT_TIMEOUT
+
+    @staticmethod
+    def get_discovery_targets(targetip: str = "255.255.255.255"):
+        """Return a list of discovery targets for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("get_discovery_targets should be overridden")
+
+    @staticmethod
+    def get_discovery_payload():
+        """Get discovery payload for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("get_discovery_payload should be overridden")
+
+    @staticmethod
+    def try_get_discovery_info(port, data):
+        """Try get the discovery info from supplied data after a discovery broadcast has resulted in a received datagram."""
+        raise SmartDeviceException("try_get_discovery_info should be overridden")
+
+    async def try_query_discovery_info(self):
+        """Try to query a device for discovery info."""
+        raise SmartDeviceException("try_query_discovery_info should be overridden")
+
+    async def query(self, request: Union[str, Dict], retry_count: int = 3) -> Dict:
+        """Query the device for the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("query should be overridden")
+
+    async def close(self) -> None:
+        """Close the protocol.  Abstract method to be overriden."""
+        raise SmartDeviceException("close should be overridden")
```
This can be `abs.abstractmethod` instead

sdb9696 replied:
Will do

bdraco commented on kasa/klapprotocol.py:
```diff
@@ -0,0 +1,512 @@
+
+import aiohttp
+
+# pycryptodome
+from Crypto import Random
+from Crypto.Cipher import AES
+from Crypto.Util import Padding
```
Home Assistant uses cryptography instead of pycryptdome for most libraries

https://cryptography.io/en/latest/hazmat/primitives/symmetric-encryption/

sdb9696 replied:
Now updated to use cryptography

rytilahti commented on kasa/klapprotocol.py:
```diff
@@ -0,0 +1,512 @@
+    """
+
+    DEFAULT_PORT = 80
+    DISCOVERY_PORT = 20002
+    DISCOVERY_TARGET = ("255,255,255,255", DISCOVERY_PORT)
+    DISCOVERY_BROADCAST_PAYLOAD = binascii.unhexlify("020000010000000000000000463cb5d3")
+    DISCOVERY_QUERY = {"system": {"get_sysinfo": None}}
```
Let's keep the discovery parts inside the `discover.py`. I bought a tapo device and looks like it also supports this discovery mode, so we could inform users about such devices and advice them to look for using tapo library for comms:
```
DEBUG:kasa.discover:Got {'error_code': 0,
 'result': {'device_id': 'xxxx',
            'device_model': 'P110(EU)',
            'device_type': 'SMART.TAPOPLUG',
            'factory_default': False,
            'ip': '192.168.xxx',
            'is_support_iot_cloud': True,
            'mac': '48-22-54-xx',
            'mgt_encrypt_schm': {'encrypt_type': 'AES',
                                 'http_port': 80,
                                 'is_support_https': False,
                                 'lv': 2},
            'obd_src': 'tplink',
            'owner': 'xxxx'}}
```

The handshake itself looks different (https://github.com/fishbigger/TapoP100/blob/main/PyP100/PyP100.py#L113) and I'm just getting this as a response for `/app/handshake1` when trying to query the device:
```
HTTP/1.1 200 OK
Server: SHIP 2.0
Content-Length: 49
Content-Type: text/html

<html><body><center>200 OK</center></body></html>
```

sdb9696 replied:
This should be ok now

rytilahti commented on kasa/klapprotocol.py:
```diff
@@ -0,0 +1,512 @@
+        self.remote_seed = response_data[0:16]
+        server_hash = response_data[16:]
+
+        cookie = self.jar.filter_cookies(url).get(self.TP_SESSION_COOKIE_NAME)
+        _LOGGER.debug(
+            f"Handshake1 posted at {datetime.datetime.now()}.  Host is {self.host}, Session cookie is {cookie}, Response status is {response_status}, Request was {self.local_auth_hash.hex()}"
+        )
```
Does the device set the cookie always? The same cookie name is used by tapo devices, but as the request on P110 did not return a cookie, maybe this should raise an early exception if no cookie is being set?

sdb9696 replied:
As far as I am aware the kasa klap devices always set the cookie after handshake1.

rytilahti reviewed (commented):
Hi @sdb9696 and thanks for the updated PR!

I am currently not able to go through all of this PR given its size and as I have no way to test it myself properly, but I did a partial (i.e., without klapprotocol.py) review to give some feedback and improvement ideas.

Some thoughts without specific order of importance:

- Please keep this PR as simple as possible. This means that avoid introducing any unrelated changes and create separate PRs for those, I added some comments inline.

- The existing functionality shall not be broken. We want to keep the existing APIs functioning without any changes necessary for downstreams who are already happy.

- Avoid any type of protocol-specific handling in the main classes (this includes things like conditionals, checking for instance types/ports) wherever possible. Looks like the TAPO devices are using the same discovery protocol, but other parts are different. Keeping things uncoupled makes it easier to introduce support for those devices if wanted.

- Please rebase your changes on top of the previously existing PR, as this will keep the commit history intact and give credit also to those who worked on this earlier.

My suggestion is the following:

- Extract the discovery (and other unrelated changes as commented inline) from this PR into their own so we can review and merge them separately. Starting with discovery feels like a good place to start as having available as a first step will let us inform users about unsupported devices and avoid issue reports on them.

- The discovery should be unauthenticated at first, we can revisit this when we find a proper level of abstraction. Instead of having a separate "unauthenticateddevice", maybe we can find a solution where trying to access authenticated features would raise an exception and allow authentication?

- I will see if I find some time to look into the APIs at some point, maybe I will try integrating the tapo protocol for getting a feeling on how the relevant architectural parts should be done, but I will make no promises on that.


P.S. Are devices using this protocol still being produced and available? That is, could I go and buy myself one for testing (e.g. https://www.amazon.co.uk/TP-Link-Monitoring-SmartThings-Wireless-KP115/dp/B08LZWBTR6)? I am just trying to get an idea if it would be a better idea to have a separate fork/project for this product line instead of trying to shoehorn support into this project? AFAIK, this protocol has never been seen in the wild in EU and US versions of kasa devices, and probably never will considering that plenty of new devices are using tapo branding now...

sdb9696 commented:
Hi @rytilahti, thanks for taking the time to review.  Some comments inline

> Hi @sdb9696 and thanks for the updated PR!
> 
> I am currently not able to go through all of this PR given its size and as I have no way to test it myself properly, but I did a partial (i.e., without klapprotocol.py) review to give some feedback and improvement ideas.
> 
> Some thoughts without specific order of importance:
> 
> * Please keep this PR as simple as possible. This means that avoid introducing any unrelated changes and create separate PRs for those, I added some comments inline.

Understood, will do.

> * The existing functionality shall not be broken. We want to keep the existing APIs functioning without any changes necessary for downstreams who are already happy.

Understood, will do, although I'd appreciate it if you could reconsider adding "port" to all the device classes that was introduced in PR https://github.com/python-kasa/python-kasa/pull/471 as it's quite contrary to the idea of having different protocols.  Once the protocol is also added to the device constructor the only purpose of having the port is to display it in the  only added in order to display the port in the state() function in cli.py.  I'd be happy to do a separate PR simply to replace port with protocol for the device classes for the next release.

> * Avoid any type of protocol-specific handling in the main classes (this includes things like conditionals, checking for instance types/ports) wherever possible. Looks like the TAPO devices are using the same discovery protocol, but other parts are different. Keeping things uncoupled makes it easier to introduce support for those devices if wanted.

I tried to take on board your comments from the original PR for this around moving conditional logic out of discovery and into the protocol classes but it's difficult to completely avoid the authenticated/unauthenticated flow here.  The reason is there are  different flows due to the way it's been implemented by tp-link and the very real possibility of not being able to authenticate.  Broadcast discover() results in some basic info being returned, you then have to do a separate get_info call which could succeed or fail.  Then there is discover_single which is simple if authentication succeeds but if it fails a subsequent broadcast is needed to get that basic info.  I'll try to think of how I can further extract this away to reduce the impact on the discovery class but hopefully you can see it's quite intrinsic.

> * Please rebase your changes on top of the previously existing PR, as this will keep the commit history intact and give credit also to those who worked on this earlier.
> 
I will try to do this

> My suggestion is the following:
> 
> * Extract the discovery (and other unrelated changes as commented inline) from this PR into their own so we can review and merge them separately. Starting with discovery feels like a good place to start as having available as a first step will let us inform users about unsupported devices and avoid issue reports on them.

Ok will do

> * The discovery should be unauthenticated at first, we can revisit this when we find a proper level of abstraction. Instead of having a separate "unauthenticateddevice", maybe we can find a solution where trying to access authenticated features would raise an exception and allow authentication?

Not sure how to this easily for a few following reasons:  We don't know what Device type to create if we can't authenticate.  We could try to derive it from the discovery info payload but that would make Discovery protocol aware and you're then left with a device that errors all the time.  In my homeassistant fork I create a device with no entities when encountering this UnauthenticatedDevice which works well as there's no functionality to be expected but if it's treated as a normal device homeassistant would produce errors.  It is actually common for my devices to suddenly be unable to authenticate as the kasa app seems to update the device owner when it changes state on a schedule and you need to manually switch the light on and off via the app to get the right owner back.  

> * I will see if I find some time to look into the APIs at some point, maybe I will try integrating the tapo protocol for getting a feeling on how the relevant architectural parts should be done, but I will make no promises on that.
> 
> P.S. Are devices using this protocol still being produced and available? That is, could I go and buy myself one for testing (e.g. https://www.amazon.co.uk/TP-Link-Monitoring-SmartThings-Wireless-KP115/dp/B08LZWBTR6)? I am just trying to get an idea if it would be a better idea to have a separate fork/project for this product line instead of trying to shoehorn support into this project? AFAIK, this protocol has never been seen in the wild in EU and US versions of kasa devices, and probably never will considering that plenty of new devices are using tapo branding now...

Good question, I would say it's hard to get hold of these devices new now but I know there are a lot of people who have these and really want to the get them working with homeasistant.  Also a big driver for me creating a base class of TPLinkAuthenticationProtocol was this comment by TP-link https://community.tp-link.com/en/smart-home/forum/topic/239364?page=1 that they would in the future be updating all kasa devices to use a more secure local protocol.  The trouble with a separate fork is that it would be great to get this working out of the box with homeassistant.  Let me know what you want to do  though before I put in all the effort above :)

sdb9696 commented on .github/workflows/ci.yml:
```diff
@@ -96,16 +96,22 @@ jobs:
       - uses: "actions/setup-python@v2"
         with:
           python-version: "${{ matrix.python-version }}"
+      - name: "Install dependencies pypy windows-latest"
```
Will do

sdb9696 commented on .github/workflows/codeql-analysis.yml:
```diff
@@ -33,3 +33,19 @@ jobs:
+      uses: advanced-security/filter-sarif@v1
+      with:
+        patterns: |
+          -**/kasa/klapprotocol.py:py/weak-sensitive-data-hashing
```
Sadly I believe this the only way which is why this tool exists :(

sdb9696 replied:
Seems the codeql excepion is no longer being raised, perhaps after the move to cryptography, so have now removed this

sdb9696 commented on kasa/auth.py:
```diff
@@ -0,0 +1,45 @@
+        self.password = password
+
+
+class TPLinkAuthProtocol(TPLinkProtocol):
+    """Base class for authenticating protocol."""
+
+    def __init__(
+        self,
+        host: str,
+        port: int = 0,
+        auth_credentials: Optional[AuthCredentials] = AuthCredentials(),
+    ):
+        super().__init__(host=host, port=port)
+
+        self.auth_credentials: AuthCredentials
+        if auth_credentials is None:
+            self.auth_credentials = AuthCredentials()
+        else:
+            self.auth_credentials = auth_credentials
+        self._authentication_failed = False
+
+    @property
+    def authentication_failed(self):
+        """Will be true if authentication negotiated but failed, false otherwise."""
+        return self._authentication_failed
+
+    @authentication_failed.setter
+    def authentication_failed(self, value):
+        self._authentication_failed = value
+
+    def parse_unauthenticated_info(self, unauthenticated_info) -> Dict[str, str]:
+        """Return parsed unauthenticated info for the protocol."""
+        raise SmartDeviceException("parse_unauthenticated_info should be overridden")
```
I added this for two reasons, firstly kasa's statement that they would be adding a new strong local api to all devices at some point in the future https://community.tp-link.com/en/smart-home/forum/topic/239364?page=1 and secondly so that discovery logic would be "Authentication" aware rather than klap aware.  Happy to remove if you think it's unnecessary

sdb9696 replied:
This is now removed from the PR

sdb9696 commented on kasa/cli.py:
```diff
@@ -129,9 +130,39 @@ def _device_to_serializable(val: SmartDevice):
+    required=False,
+    help="Password to use to authenticate to device.",
+)
+@click.option(
+    "--discoverytimeout",
+    default=3,
+    required=False,
+    help="Timeout for discovery.",
+)
```
Will do

sdb9696 commented on kasa/cli.py:
```diff
@@ -177,15 +208,26 @@ def _nop_echo(*args, **kwargs):
             echo(f"No device with name {alias} found")
             return
 
+    if password != "" and username == "":
+        click.echo("Using a password requires a username")
+        return
+
+    if username != "":
+        authentication = AuthCredentials(username=username, password=password)
+    else:
+        authentication = AuthCredentials()
+
```
Ok

sdb9696 commented on kasa/cli.py:
```diff
@@ -281,7 +334,7 @@ async def state(dev: SmartDevice):
     echo(f"[bold]== {dev.alias} - {dev.model} ==[/bold]")
     echo(f"\tHost: {dev.host}")
-    echo(f"\tPort: {dev.port}")
+    echo(f"\tPort: {dev.protocol.port}")
```
See my bigger comment on this.  dev.port is redundant along with my changes as the device doesn't need to know it's port, it's in the protocol and only used in the line above.  Happy to create a separate PR to replace port with protocol prior to the next release.  That would also help make this PR smaller.

sdb9696 commented on kasa/discover.py:
```diff
@@ -33,15 +34,21 @@ def __init__(
         *,
         on_discovered: Optional[OnDiscoveredCallable] = None,
-        target: str = "255.255.255.255",
+        targetip: str = "255.255.255.255",
```
Ok will revert

sdb9696 commented on kasa/discover.py:
```diff
@@ -64,28 +71,77 @@ def connection_made(self, transport) -> None:
+            # If a discover_single is attempted for an device requiring authentication and authentication fails
+            # then it will set self.protocol and trigger a datagram discovery to get other useful info
+            if self.protocol is None or isinstance(self.protocol, proto_class):
+                for target in proto_class.get_discovery_targets(self.targetip):
```
Ok, will update

sdb9696 commented on kasa/discover.py:
```diff
@@ -64,28 +71,77 @@ def connection_made(self, transport) -> None:
-        _LOGGER.debug("[DISCOVERY] %s << %s", ip, info)
+        info = None
+        for proto_class in TPLinkProtocolConfig.enabled_protocols():
+            # If a discover_single is attempted for an device requiring authentication and authentication fails
+            # then it will set self.protocol and trigger a datagram discovery to get other useful info
```
See my bigger comment on this, I will think about it some more though

sdb9696 commented on kasa/exceptions.py:
```diff
@@ -3,3 +3,7 @@
     """Base exception for device errors."""
+
+
+class SmartDeviceAuthenticationException(Exception):
```
I believe I always catch this exception and behave accordingly, i.e. device is in an unauthenticated state.  I'll check this again

sdb9696 commented on kasa/protocolconfig.py:
```diff
@@ -0,0 +1,15 @@
+"""Module containing the configured kasa protocols."""
```
Are you ok if I just put the protocols list in the discovery class then?

sdb9696 replied:
This is now removed

sdb9696 commented on kasa/tests/fixtures/HS100(UK)_4.1_1.1.0.json:
```diff
@@ -0,0 +1,44 @@
+{
```
Ok

sdb9696 commented on kasa/tests/test_cli.py:
```diff
@@ -108,4 +116,7 @@ async def test_json_output(dev: SmartDevice, mocker):
     assert res.exit_code == 0
-    assert json.loads(res.output) == dev.internal_state
+
+    res_json = json.loads(res.output)
+
+    assert res_json == dev.internal_state
```
Ok

sdb9696 commented on kasa/unauthenticateddevice.py:
```diff
@@ -0,0 +1,238 @@
+"""Module for smart plugs (HS100, HS110, ..)."""
```
I don't think so because it is possible and actually happens that you end up with a device that can't authenticate and need the user to be aware of that.  If you look at line 156 in __init__.py of my homeassistant changes to support you can see I create a device with no entities when authentication fails: https://github.com/sdb9696/core/pull/1/files

sdb9696 commented on pyproject.toml:
```diff
@@ -28,14 +28,20 @@ pydantic = "^1"
+pyyaml = { version = ">=6.0.1", optional = true }
 
-[tool.poetry.dev-dependencies]
+[tool.poetry.group.dev.dependencies]
```
Will do

sdb9696 commented on kasa/protocol.py:
```diff
@@ -25,23 +25,102 @@
 
     INITIALIZATION_VECTOR = 171
     DEFAULT_PORT = 9999
-    DEFAULT_TIMEOUT = 5
```
Fixed

shuanang commented:
Hello, I'm currently using your commit "Remove whitespace changes and re-fix codeql". I used the following command to toggle my HS100 V4.1 firmware V1.1.0 on and off:

`kasa --enable_klap --username XXXXX@hotmail.com --password XXXXXX --host 192.168.1.111 off`
where I have set a static IP 192.168.1.111 for the HS100. 

It worked fine for a 5-6 hours in a script that runs every few minutes, and suddenly it threw an error
`Got error: SmartDeviceAuthenticationException("Server response doesn't match our challenge on ip 192.168.1.111") `
which I'm unable to resolve. The app works fine as usual. 

Any advice would be helpful, thanks!

mthole commented:
I ordered a box of KP125M's this week and quickly ended up here when I couldn't get them into HomeAssistant. Appreciate you all working on the updates here! 🎉

mthole commented:
@shuanang I'm getting a similar result with my KP125M on the [same commit](https://github.com/python-kasa/python-kasa/commit/99073f714b968b44dde46d70a848560880778a5b).

```
➜  python-kasa git:(master) poetry run kasa --enable_klap --username XXX@gmail.com --password XXX --host 192.168.0.63            
No --type defined, discovering..
Got error: SmartDeviceAuthenticationException("Server response doesn't match our challenge on ip 192.168.0.63")
```

If I pass the `--type plug`, it seems to try to connect over 9999, which if I understand correctly, is the old/unsupported mechanism.

```
➜  python-kasa git:(master) poetry run kasa --enable_klap --username XXX@gmail.com --password XXX --host 192.168.0.63 --type plug
Got error: SmartDeviceException("Unable to connect to the device: 192.168.0.63:9999:  Connect call failed ('192.168.0.63', 9999)")
```

sdb9696 commented:
> Update: The same command works again, after waiting ~2 hours. Could it be their server limiting my requests?
> 
> Hello, I'm currently using your commit "Remove whitespace changes and re-fix codeql" (commit [99073f7](https://github.com/python-kasa/python-kasa/commit/99073f714b968b44dde46d70a848560880778a5b)). I used the following command to toggle my HS100 V4.1 firmware V1.1.0 on and off:
> 
> `kasa --enable_klap --username XXXXX@hotmail.com --password XXXXXX --host 192.168.1.111 off` where I have set a static IP 192.168.1.111 for the HS100.
> 
> It worked fine for a 5-6 hours in a script that runs every few minutes, and suddenly it threw an error `Got error: SmartDeviceAuthenticationException("Server response doesn't match our challenge on ip 192.168.1.111") ` which I'm unable to resolve. The app works fine as usual.
> 
> Any advice would be helpful, thanks!

Hi, apologies for the delay in replying I have been away.  I think the issue you have been encountering is the same one I have been getting and it seems to be related to the device randomly changing owner.  It seems to be something to do with when the device operates on it's internal scheduler and it changes the owner to something that produced this md5 hash: 994661e5222b8e5e3e1d90e73a322315.  

I was wrestling with this before I went away but I've now been able to identify the new owner from the kasa apk source code as kasa@tp-link.net.  I've updated the PR to fall back and try these credentials if the correct username and password does not work.  This seems to be working fine for me now.  It would great if you could try out with my latest commit and let me know.

sdb9696 commented:
> @shuanang I'm getting a similar result with my KP125M on the [same commit](https://github.com/python-kasa/python-kasa/commit/99073f714b968b44dde46d70a848560880778a5b). KP125M, firmware 1.1.0
> 

I don't think your issue is the same as @shuanang.  Did this ever work for you?  (@shuanang had it working then it stopped).  afaik this fix will not work with the KP125M as it only affects HS100 with the new klap protocol.  It looks to me like your KP125M might be doing something with the TAPO protocol.

rytilahti commented:
@mthole feel free to give #499 (it implements tapo support using an external lib) a try and let us know if that works on that KP125M. Alternatively, you could also try #488 to see and let us know how the discovery response looks for this device so we can get an idea if we need to accommodate those separately.

shuanang commented:
> > Update: The same command works again, after waiting ~2 hours. Could it be their server limiting my requests?
> > Hello, I'm currently using your commit "Remove whitespace changes and re-fix codeql" (commit [99073f7](https://github.com/python-kasa/python-kasa/commit/99073f714b968b44dde46d70a848560880778a5b)). I used the following command to toggle my HS100 V4.1 firmware V1.1.0 on and off:
> > `kasa --enable_klap --username XXXXX@hotmail.com --password XXXXXX --host 192.168.1.111 off` where I have set a static IP 192.168.1.111 for the HS100.
> > It worked fine for a 5-6 hours in a script that runs every few minutes, and suddenly it threw an error `Got error: SmartDeviceAuthenticationException("Server response doesn't match our challenge on ip 192.168.1.111") ` which I'm unable to resolve. The app works fine as usual.
> > Any advice would be helpful, thanks!
> 
> Hi, apologies for the delay in replying I have been away. I think the issue you have been encountering is the same one I have been getting and it seems to be related to the device randomly changing owner. It seems to be something to do with when the device operates on it's internal scheduler and it changes the owner to something that produced this md5 hash: 994661e5222b8e5e3e1d90e73a322315.
> 
> I was wrestling with this before I went away but I've now been able to identify the new owner from the kasa apk source code as [kasa@tp-link.net](mailto:kasa@tp-link.net). I've updated the PR to fall back and try these credentials if the correct username and password does not work. This seems to be working fine for me now. It would great if you could try out with my latest commit and let me know.

Hey, thanks for getting back! I'd like to update that I've tried your latest commit [cce1f6e](https://github.com/sdb9696/python-kasa/commit/cce1f6ee3535e4e32aa338338d1fb69cabf92890) and my script has been running without issues for almost a day now. Thanks for the hard work and for detailing the issue! :-)

petretiandrea commented:
Trying similar code to my Tapo L530, the hanshake1 doesn't match. By doing some reverse engineering to Tapo App I've found a solution: you must use `remote_seed` when calculate the local hash which must be match with server hash. This is a snippet of my code:
```python
local_seed_auth_hash = KlapProtocol._sha256(
    self._local_seed + remote_seed + self.local_auth_hash
)
```

sdb9696 commented:
Interesting.  @petretiandrea could you post what the discovery response looks like for a device running this new firmware?  Does it only affect the L530?  I bought a P110 the other day so could test with that if you think it will pick up the new firmware.

petretiandrea commented:
> Interesting. @petretiandrea could you post what the discovery response looks like for a device running this new firmware? Does it only affect the L530? I bought a P110 the other day so could test with that if you think it will pick up the new firmware.

Authentication failed for device:
` {'result': {'device_id': '6dc952ccdc764f54faf415bac99886a6', 'owner': '006EE6989D4F3A0F47715B4F1585CC27', 'device_type': 'SMART.TAPOBULB', 'device_model': 'L530E(EU)', 'ip': '192.168.1.17', 'mac': '54-AF-97-B2-5C-18', 'is_support_iot_cloud': True, 'obd_src': 'tplink', 'factory_default': False, 'mgt_encrypt_schm': {'is_support_https': False, 'encrypt_type': 'KLAP', 'http_port': 80, 'lv': 2}}, 'error_code': 0}`

sdb9696 commented:
>` {'result': {'device_id': '****', 'owner': '****', 'device_type': 'SMART.TAPOBULB', 'device_model': 'L530E(EU)', 'ip': '****', 'mac': '*****', 'is_support_iot_cloud': True, 'obd_src': 'tplink', 'factory_default': False, 'mgt_encrypt_schm': {'is_support_https': False, 'encrypt_type': 'KLAP', 'http_port': 80, 'lv': 2}}, 'error_code': 0}`

So the discovery response for a kasa klap device doesn't have `'lv: 2` so this probably can be used to decide whether to include both seeds as per your previous post.  kasa klap:

`{'result': {'ip': '192.168.123.123', 'mac': '12-34-56-78-9A-BC', 'device_id': '123456789ABCDEF123456789ABCDEF12', 'owner': 'fedcba987654321fedcba987654321fe', 'device_type': 'IOT.SMARTPLUGSWITCH', 'device_model': 'HS100(UK)', 'hw_ver': '4.1', 'factory_default': True, 'mgt_encrypt_schm': {'is_support_https': False, 'encrypt_type': 'KLAP', 'http_port': 80}}, 'error_code': 0}`

> Authentication failed for device: 

Where did this come from?  Did you get the discovery response via udp on port 20002 or was "Authentication failed" returned somehow along with the `{result}` due to an authentication failure?  It seemed from your previous post that you have managed to get klap working with both seeds so just trying to understand here.  Also is the encryption for TAPO klap the same as in the klapprotocol.py file or is it also different like the handshake?

sdb9696 commented:
> `Authentication failed for device:` is print by python-kasa by launching discovery using the code of this PR

 Got it thanks

> EDIT: you can found my implementation (inspired by this PR) here: https://github.com/petretiandrea/plugp100/blob/main/plugp100/protocol/klap_protocol.py. This library is used to integrate Tapo devices with home assistant. The chiper is the same implemented in this PR

That’s really good and glad to see klapprotocol.py in use. You turned around this fix really fast so I’m sure your users are v.grateful. At some point we should compare notes on how this new protocol behaves in the real world.

sdb9696 commented:
Hi @rytilahti, I've rebased this PR from master following your merge of the discovery changes in https://github.com/python-kasa/python-kasa/pull/488.  I've also incorporated the comments from https://github.com/python-kasa/python-kasa/pull/488#issuecomment-1684261410 with regard to the protocol not being in the constructors.  Let me know what you think so far.

rytilahti reviewed (commented):
See some comments inline, I didn't yet take a look into the klapprotocol implementation but I think the best steps forward would be to split this PR even more:

1. A PR to add support for credentials (incl. the new exception), this will unblock me so I can experiment with the tapo support.
    * This involves adding `authentication` to the ctor of `SmartDevice` derived classes.
    * `SmartDevice.__init__` can store this info, which can then be used wherever needed.

2. Refactor the discovery support to make it easier to write tests. Instead of having everything contained in the datagram received, let's split it up:

    * `_get_device_instance_legacy(info)` is responsible for constructing a device instance for the old 9999/udp responses.
   * `_get_device_instance(info, credentials)` which handles the 20002/udp initialization. I'm not sure if this should already do the update try, but rather leave it for the first time the downstream calls `update()` (which would then raise the exception if auth fails).
        * This method is also responsible for setting the protocol class for klap devices.
        * This would be the place where I could hook the tapo plug draft nicely, and in case tplink decides to introduce klap for future tapo devices, this library could be extended to support those directly.

I think there are still some open questions on how to approach integrating this with homeassistant, but I think we are getting there :-)

Open questions include things like:
* What if discovery was done using invalid creds or no creds at all? I suppose that could be a fairly normal downstream use-case, so I'm somehow imagining that initialization of device instances should work without valid creds, `update()` raising the exception if necessary.
* On that topic, maybe `Discover.discover` should not have a separate "auth failed", but deliver those devices just as-is, and leave it to the downstream user to 1) set new Credentials/Authenticaton object and 2) retry `update()`?

Do you have a homeassistant integration fork using this PR for testing? I could experiment with that and the tapo plug to get a better feeling about how it would work with homeassistant, at the moment I'm reading the creds from the envvars in that tapo draft PR.

rytilahti commented on kasa/discover.py:
```diff
@@ -128,6 +169,10 @@ def datagram_received(self, data, addr) -> None:
         device = device_class(ip, port=port)
         device.update_from_discover_info(info)
 
+        # Override protocol if klap created and used succesfully
+        if protocol:
+            device.override_protocol(protocol)
```
This should be done in the instance initialization method where needed. Instead of having a separate `override_protocol`, I'd simply set the instance variable for the time being.

sdb9696 replied:
Ok

rytilahti commented on kasa/discover.py:
```diff
@@ -88,29 +97,61 @@ def do_discover(self) -> None:
+                self.unsupported_devices[ip] = info
+                if self.on_unsupported is not None:
+                    asyncio.ensure_future(self.on_unsupported(info))
+                if self.discovered_event is not None and "255" not in self.target[
+                    0
+                ].split("."):
+                    self.discovered_event.set()
+                return
```
Is this really necessary as it reads rather hacky?

sdb9696 replied:
It's consistent with how I did it in the add new discovery protocol PR which has this check in a few places.  The thinking behind it is that the discovery done event is only triggered when targeting a host rather than a broadcast.

sdb9696 commented on kasa/cli.py:
```diff
@@ -195,6 +230,14 @@ def _nop_echo(*args, **kwargs):
+        click.echo("Using a password requires a username")
+        return
+
+    auth_credentials: Optional[AuthCredentials] = (
+        AuthCredentials(username=username, password=password) if enable_klap else None
+    )
```
Done in PR https://github.com/python-kasa/python-kasa/pull/507

sdb9696 commented on kasa/auth.py:
```diff
@@ -0,0 +1,9 @@
+"""Authentication class for username / passwords."""
+
+
+class AuthCredentials:
```
I think the default should be empty string rather than none.  Apparently kasa klap devices disconnected from the cloud have an empty string username and password.  Better to try to authenticate with this every time and report failures than have to handle the meaning of None values for username or password no?

rytilahti replied:
This fallback should be done in the klapprotocol class itself, having it default to `None` makes it clear that it's undefined rather than something the user inputted.

sdb9696 commented on kasa/cli.py:
```diff
@@ -137,6 +138,36 @@ def _device_to_serializable(val: SmartDevice):
+)
+@click.option(
+    "--username",
+    default="",
```
See previous comment

sdb9696 commented on kasa/cli.py:
```diff
@@ -137,6 +138,36 @@ def _device_to_serializable(val: SmartDevice):
+)
+@click.option(
+    "--password",
+    default="",
```
See previous comment

sdb9696 commented on kasa/cli.py:
```diff
@@ -195,6 +230,14 @@ def _nop_echo(*args, **kwargs):
             echo(f"No device with name {alias} found")
             return
 
+    if password != "" and username == "":
+        click.echo("Using a password requires a username")
+        return
```
I've removed this entirely as it's a bit of a spurious check and an authentication failure should suffice

sdb9696 commented on kasa/cli.py:
```diff
@@ -137,6 +138,36 @@ def _device_to_serializable(val: SmartDevice):
     required=False,
     help="Timeout for discovery.",
 )
+@click.option(
+    "--enable-klap",
+    envvar="KASA_ENABLE_KLAP",
+    required=False,
+    default=False,
+    is_flag=True,
+    help="Set to true to enable the klap protocol",
+)
```
Ok

sdb9696 commented on kasa/cli.py:
```diff
@@ -137,6 +138,36 @@ def _device_to_serializable(val: SmartDevice):
+    envvar="KASA_CLOUD_PASSWORD",
+    help="Password to use to authenticate to device.",
+)
+@click.option(
+    "--show-authentication-failed",
+    envvar="KASA_SHOW_AUTHENTICATION_FAILED",
+    required=False,
+    default=False,
+    is_flag=True,
+    help="Print out discovered devices that failed to authenticate",
+)
```
Ok

rytilahti commented on kasa/cli.py:
```diff
@@ -203,7 +246,9 @@ def _nop_echo(*args, **kwargs):
         dev = TYPE_TO_CLASS[type](host)
```
This is one of the reasons, yes. I think it makes the API clear across all different current and future protocols, as the `SmartDevice` ctor signature is always the same no matter what low-level plumbing is being used.

I haven't given it too much thought yet, but we could potentially extend `Smart*` to use the `discover_single` in the future to detect the device type and act accordingly, if need to be. This way, any existing code like the example below would keep working just as it was if this library gets extended with support for new transports or other low-level protocols:
```
from kasa import SmartDevice, Credentials

async def turnoff(dev):
    await dev.turn_off()

if __name__ == "__main__":
    creds = Credentials("foobar@example.com", "1234")
    old_plug = SmartDevice("legacy-plug", credentials=creds)
    tapo_plug = SmartDevice("tapo-plug", credentials=creds)
    klap_plug = SmartDevice("klap hs110", credentials=creds)
    asyncio.run(asyncio.gather(old_plug, tapo_plug, klap_plug))
```

rytilahti commented:
> Done in PR #507 but only as far as discovery as I still think the credentials belong to the protocol rather than the device. Perhaps only the TAPO shim can add it to it's own `__init__`?

The thing is (as mentioned in a comment above) is that it is better to have a common, stable interface for downstream users no matter what device they may have.

> I do but it's not great right now as I was trying to store the credentials globally in a hacky way. I'm planning to update ti store the credentials in an OptionFlow config and then update all the configentries whenever one of the OptionsFlow changes. Unless you have a better suggestion. Once you've merged the credentials PR I'll update it work with the new classname and point you to it

I don't know if there's a nicer way to store the creds for the integration, but I hope so. Updating config entries by force has a couple of problems:
* What if the user has different devices using different creds?
* What if some of the devices are blocked from the internet and the account password gets changed, i.e., the device internal secrets are only updated to some of the devices?

sdb9696 commented on kasa/discover.py:
```diff
@@ -310,3 +372,13 @@ def _get_device_class(info: dict) -> Type[SmartDevice]:
+
+    @staticmethod
+    def _is_klap_discovery_info(info: dict) -> bool:
+        return (
```
Now implemented as per your comment

sdb9696 commented:
>>> Btw, do you have a homeassistant integration fork using this PR for testing? I could experiment with that and the tapo plug to >>> get a better feeling about how it would work with homeassistant, at the moment I'm reading the creds from the envvars in that tapo draft PR.

>> I do but it's not great right now as I was trying to store the credentials globally in a hacky way. I'm planning to update ti store the credentials in an OptionFlow config and then update all the configentries whenever one of the OptionsFlow changes. Unless you have a better suggestion. Once you've merged the credentials PR I'll update it work with the new classname and point you to it
>I don't know if there's a nicer way to store the creds for the integration, but I hope so. Updating config entries by force has a couple of problems:

>What if the user has different devices using different creds?
>What if some of the devices are blocked from the internet and the account password gets changed, i.e., the device internal >secrets are only updated to some of the devices?

Have a look at PR https://github.com/sdb9696/core/pull/1 for a working ha fork with this PR.  I am using the options flow now as I think that is the best way to go and the user can configure whether they want to sync credentials across other config entries or not

sdb9696 commented:
Hi @rytilahti, looking forwards to getting this one closed out now.  Two things:

- There is quite a long conversation on this PR, a lot of which is now addressed in the other PRs we finished.  Let me know if you want me to create a fresh PR so we can have a smaller conversation
- The codeQL check for weak hashing due to the md5 hash is constantly failing on this PR.  I think if you dismiss the alert in github it will stop failing the check and we won't have to configure CodeQL to exclude it or do any post processing

Thanks!

rytilahti commented:
Hi @sdb9696 and thanks for your efforts to get this properly integrated, and opening the potential for future extensions! :-)

>  There is quite a long conversation on this PR, a lot of which is now addressed in the other PRs we finished.  Let me know if you want me to create a fresh PR so we can have a smaller conversation

Your call, it probably makes sense to start with a clean PR at this point if it's not too complicated while still keeping the git history intact?

>  The codeQL check for weak hashing due to the md5 hash is constantly failing on this PR.  I think if you dismiss the alert in github it will stop failing the check and we won't have to configure CodeQL to exclude it or do any post processing

I can definitely do that. And looks like it's working as expected with manual dismissal as it's green again.

rytilahti commented on kasa/smartdevice.py:
```diff
@@ -205,15 +209,19 @@ def __init__(
         #       are not accessed incorrectly.
         self._last_update: Any = None
+
+        # Homeassistant uses the mac for storing devices and alias for display so this property allows updating of
+        # some properties from the new discovery info even if the device can't authenticate
+        self._requires_update_overrides: dict = {}
```
Just a quick note, this is something that is currently done (partially, at least) using the `update_from_discover_info` method.
